### ISSUE TITLE
Data parsing bug of digits saved as text and the possible fix, etc.

### ISSUE THREAD

**Issue by dgy18787** [cosmocode/dokuwiki-plugin-struct#665]
Hello,

Thank you guys for developing this convenient plugin. Recently I found a possible bugs of struct when it parsing digits and apply them in a drop-down filter, here's how the bug occur:

1. There's a schema, the field name is original_date, type is text, it is used to store the release year of each song. As some song is not clear in the exact release year, there will be a "?" after a year, so the data would be like 1992, 1993? or [1992, 1995](for multiple values).
2. For a quick search, I made a page using struct filter and struct table, but I found that the dropdown of "original_date" is filled with irrregular numbers (like 1978 parsed as another digits), but the digits with suffix "?" is normally parsed.

There's a possible fix for the issue, just modify meta/AggregationFilter.php, there's two place to change :

```php
foreach (array_unique($colData['values']) as $value => $displayValue) {
//$colData['values'] --> array_unique($colData['values'])

    $current = false;
    $dyn = $this->searchConfig->getDynamicParameters();
    $allFilters = $dyn->getFilters();
    if (isset($allFilters[$column->getFullQualifiedLabel()])) {
        if ($allFilters[$column->getFullQualifiedLabel()][1] == $displayValue) {
            $current = true;
        }
        $dyn->removeFilter($column); // remove previous filter for this column
    }
    if (!$current) {
        // add new filter unless it's the current item
        $dyn->addFilter($column, '=', $displayValue);
    }
    $params = $dyn->getURLParameters();
    $filter = buildURLparams($params);

    $this->renderer->doc .= '<li ' . ($current ? 'class="active"' : '') . '><div class="li">';
    $column->getType()->renderTagCloudLink($displayValue, $this->renderer, $this->mode, $this->id, $filter, 100);
//modified $displayValue

    $this->renderer->doc .= '</div></li>';
}
```

**Comment by splitbrain** [comment]
please report one issue per ticket only. I removed the additional points.

**Comment by splitbrain** [comment]
> I found that the dropdown of "original_date" is filled with irrregular numbers (like 1978 parsed as another digits), but the digits with suffix "?" is normally parsed.

I don't understand this sentence. Can you please try again in describing the actual problem? Eg, what is the exact data in your table, what syntax are you using, what are you expecting and what are you seeing instead. Add screenshots if helpful.

**Comment by dgy18787** [comment]
Sorry for the late response, yes, I'm trying to explain the details. I have updated the struct to the latest version, the problem is as follows:

1. in my table there is a "release date" column for each pageid (which represent a keyboard model). As release date of some model are unknown or inaccurate, some times we put a "?" behind the year, so the data type of this column is set to "Text". The table looks like:

![bug1](https://github.com/cosmocode/dokuwiki-plugin-struct/assets/2542209/a4ea807c-f35b-4f8c-918e-eb11c66cfdbf)

There're three types of content in this column, the accurate year (XXXX in four digits), the inaccurate year (XXXX?) and unknown year (only a question mark "?").

2. As I try to complete aggregation and add the filter, I found that the list item under the release_date dropdown doesn't show the correct items, like the four digits is parsed into other numbers(which is wrong), but "?" and "XXXX?" is correctly parsed, which looks like:

![bug2](https://github.com/cosmocode/dokuwiki-plugin-struct/assets/2542209/e2630b70-9a5a-4d31-bc3a-9f42d971d52e)

There's no three digits in this column of the schema, so I suppose this might be a bug. 

The page screenshot is from https://demodb.org/search/model_result, I now use the solution I mentioned before to solve this temporarily.
